core/primitives/src/block_header.rs

```diff
@@ -0,0 +1,481 @@
+}
+
+/// Block approval by other block producers.
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, PartialEq, Eq)]
+pub struct ApprovalMessage {
+    pub approval: Approval,
+    pub target: AccountId,
+}
+
+impl ApprovalInner {
+    pub fn new(
+        parent_hash: &CryptoHash,
+        parent_height: BlockHeight,
+        target_height: BlockHeight,
+    ) -> Self {
+        if target_height == parent_height + 1 {
+            ApprovalInner::Endorsement(parent_hash.clone())
+        } else {
+            ApprovalInner::Skip(parent_height)
+        }
+    }
+}
+
+impl Approval {
+    pub fn new(
+        parent_hash: CryptoHash,
+        parent_height: BlockHeight,
+        target_height: BlockHeight,
+        signer: &dyn ValidatorSigner,
+    ) -> Self {
+        let inner = ApprovalInner::new(&parent_hash, parent_height, target_height);
+        let signature = signer.sign_approval(&inner, target_height);
+        Approval { inner, target_height, signature, account_id: signer.validator_id().clone() }
+    }
+
+    pub fn get_data_for_sig(inner: &ApprovalInner, target_height: BlockHeight) -> Vec<u8> {
+        [inner.try_to_vec().unwrap().as_ref(), target_height.to_be_bytes().as_ref()].concat()
+    }
+}
+
+impl ApprovalMessage {
+    pub fn new(approval: Approval, target: AccountId) -> Self {
+        ApprovalMessage { approval, target }
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, Eq, PartialEq)]
+pub struct BlockHeaderV1 {
+    pub prev_hash: CryptoHash,
+
+    /// Inner part of the block header that gets hashed, split into two parts, one that is sent
+    ///    to light clients, and the rest
+    pub inner_lite: BlockHeaderInnerLite,
+    pub inner_rest: BlockHeaderInnerRest,
+
+    /// Signature of the block producer.
+    pub signature: Signature,
+
+    /// Cached value of hash for this block.
+    #[borsh_skip]
+    pub hash: CryptoHash,
+}
+
+#[derive(BorshSerialize, BorshDeserialize, Serialize, Debug, Clone, Eq, PartialEq)]
+#[borsh_init(init)]
+pub enum BlockHeader {
+    BlockHeaderV1(BlockHeaderV1),
+}
+
+impl BlockHeader {
+    pub fn compute_inner_hash(inner_lite: &[u8], inner_rest: &[u8]) -> CryptoHash {
+        let hash_lite = hash(inner_lite);
+        let hash_rest = hash(inner_rest);
+        combine_hash(hash_lite, hash_rest)
+    }
+
+    pub fn compute_hash(prev_hash: CryptoHash, inner_lite: &[u8], inner_rest: &[u8]) -> CryptoHash {
+        let hash_inner = BlockHeader::compute_inner_hash(inner_lite, inner_rest);
+
+        return combine_hash(hash_inner, prev_hash);
+    }
+
+    pub fn init(&mut self) {
+        let hash = BlockHeader::compute_hash(
+            *self.prev_hash(),
+            &self.inner_lite_bytes(),
+            &self.inner_rest_bytes(),
+        );
+        match self {
+            BlockHeader::BlockHeaderV1(header) => header.hash = hash,
+        }
+    }
+
+    pub fn new(
+        _protocol_version: ProtocolVersion,
```

Comment:
why isn't this used?

Reply:
because right now we don't have other versions of block header.